### ISSUE TITLE
Unexpected results after using EDIT section

### ISSUE THREAD

**Issue by andrescodas** [OPM/opm-simulators#2183]
I re-wrote the SPE9 case here: https://github.com/OPM/opm-data/blob/master/spe9/SPE9_CP.DATA
in such a way that I define a bogus geometry on the GRID section, and then after I redefine all the properties on the EDIT section so that the model is equivalent to SPE9_CP again.

OPM does not seem to deliver correct results... Could you help me get it right? See the example here:
[spe9_cp_edit.tar.gz](https://github.com/OPM/opm-simulators/files/3888396/spe9_cp_edit.tar.gz)

**Comment by joakim-hove** [comment]
I am sorry but I do not think we have capacity to help you on such loose general terms. I have not looked on your deck, but apart from the possibility that you have actually done something wrong, I would like to make the following general comments:
1. The grid related keywords are not treated as the other 3D keywords - you need a grid before you can define and edit e.g. the PORO field, i.e. there is a chicken and egg problem here.

2. Opm/flow does not handle fine points with GRID versus EDIT section very well.

That said we are currently refactoring the input part of the 3D property system, and your testing might be interesting to look at in that context. Don't hold your breath though.

**Comment by andrescodas** [comment]
Hi Joakim, thanks for your prompt although not too optimistic answer.

Re 1.
In GRID I use DX DY DZ TOPS ACTNUM PERMX PERMY PERMZ PORO.  Then in EDIT I use DEPTH PORV TRANX TRANY TRANZ.  Does it exist the chicken and egg problem still?  I think the GRID is fully defined by the first set... so I don't see why the chicken and egg problem exist.

Re 2.
My guess is that the problem is in the well switching logic or in the position of the well with respect to the edited grid.  I see a lot of switching on the first iterations which does not seem correct.

I'm happy to give more details if needed!.

**Comment by joakim-hove** [comment]
Oh yes - that is certainly chicken and egg, or maybe an omelet of some sort...

The grid will be defined based on some of the keywords you mention (I do not have the ordering in front of me so I can not guarantee which set),  and the others will be ignored - EDIT or not. 

I will look into your test case, but for now I would recommend creating a normal grid and focus on the simulation.

**Comment by andrescodas** [comment]
I know you asked not to hold my breath, but I'll appreciate if you let me know if a fix is available, even if I had to compile flow myself to get it to run.

Although this example seems meaningless, I prepared it to show the failing point I'm facing in a reproducible way.  I use the related scripts in an upscaling workflow that lead to this paper https://doi.org/10.3389/fdata.2019.00033

**Comment by alfbr** [comment]
@andrescodas I am afraid most of us are pressured on time, so generally speaking it is difficult to consider problems presented with a complete data set. For instance, it is not clear if what you are asking is a feature request, bug-fix or a false-positive. If it is an easy to reproduce bug, we typically try to fix it. If it is a feature request, patches are welcome. We already have a huge back-log of sponsored feature requests that we need to address, but then please post it as a clearly formulated feature request.

**Comment by bska** [comment]
> OPM does not seem to deliver correct results

Just to add some detail to what @joakim-hove and @alfbr have already mentioned.  I ran your case in the latest development version of Flow and it produced exactly the fileset listed below.  The `.RFT` file is empty so it probably should not exist at all.  That's a bug.

In any case you mind expanding what you mean by "correct results" in the above statement?  Note that there is no `.UNRST` file at all due to your `.DATA` file not having any kind of `RPTRST` request.  Is that what you mean?

---

 * `SPE9_CP.DBG`
 * `SPE9_CP.EGRID`
 * `SPE9_CP.INIT`
 * `SPE9_CP.PRT`
 * `SPE9_CP.RFT`
 * `SPE9_CP.SMSPEC`
 * `SPE9_CP.UNSMRY`

**Comment by andrescodas** [comment]
Hi @bska and @alfbr .  Sorry for not being more specific.  with "OPM does not seem to deliver correct results" I mean that OPM and ECLIPSE generate different WOPR (and other well variables). Further, the results delivered by ECLIPSE are the same as the original SPE9.

**Comment by bska** [comment]
Quick update here.  A colleague of mine was able to identify one possible issue that might contribute to this behaviour.  Our grid processing routines appear to ignore the `DEPTH` keyword (for specifying grid block centre depths).  Your model uses that keyword and is therefore possibly affected by the missing support.

I'm looking into this more closely now.

**Comment by andrescodas** [comment]
Perfect timing @bska .  I was working on giving a follow up to this issue just now.  I'm making available the outputs I'm observing. 
![Oil_Production_Rate](https://user-images.githubusercontent.com/164282/70439363-995daf80-1a6e-11ea-8c1d-a50e365df61c.png)
[SPE9_EDIT_ECLIPSE.SMSPEC.zip](https://github.com/OPM/opm-simulators/files/3939865/SPE9_EDIT_ECLIPSE.SMSPEC.zip)
[SPE9_EDIT_OPMFLOW.SMSPEC.zip](https://github.com/OPM/opm-simulators/files/3939867/SPE9_EDIT_OPMFLOW.SMSPEC.zip)

**Comment by totto82** [comment]
You can try to make the case depth agnostic by setting the initial conditions explicitly and setting default reference depth for the wells. (It will use the depth of the top connection)

**Comment by andrescodas** [comment]
Hi @totto82, thanks for your suggestion.  I removed the DEPTH keyword and fixed the keywords DX, DY, DZ and TOPS, in such a way that I match the centroids of the cells, and the centroids of the horizontal faces to the original SPE9 model.  See the new grid:

![spe9_no_depth](https://user-images.githubusercontent.com/164282/70550646-d308e600-1b54-11ea-880f-7f87d2031e2f.png)

While this new model with now DEPTH keyword executed with Eclipse matches the previous experiment using DEPTH, the result by OPM flow is still different, but better than the previous experiment. See:
![Oil_Production_Rate_no_depth_key](https://user-images.githubusercontent.com/164282/70551097-a43f3f80-1b55-11ea-9ea5-b286e7698bf7.png)

**Comment by OPMUSER** [comment]
The difference is because the model is 2D, and has modifications in the EDIT section. If you convert the model to 3D by adding one or two slides then it should work correctly.

This is a bug.

**Comment by andrescodas** [comment]
Hi @OPMUSER .  What model is 2D?  Particularly, I think the model is 3D because the model states:

DIMENS
        24 25 15 /

Where should I add the Y-slices?

**Comment by OPMUSER** [comment]
@andrescodas my mistake, I just looked at the image above and assumed it was a 2D model, when it is actually a 3D model.

I have ran some in-house tests on other models and have only found an issue with 2D models.

The commercial simulator reports this error with your input deck:
```
 @--  ERROR  AT TIME        0.0   DAYS    ( 1-JAN-   0):
 @           TOO MUCH DATA FOUND IN SECTION    1 OF KEYWORD WELLDIMS         
 @           COMMONEST CAUSE IS MISSING / AT END OF SECTION                  
               END OF INCLUDE FILE 

 @--WARNING  AT TIME        0.0   DAYS    ( 1-JAN-   0):
 @           SPURIOUS DATA BEFORE OIL      KEYWORD                           
 @     
```
And OPM Flow gives this warning but runs:

```
Warning: Expected the kewyord: EQLDIMS to infer the number of records in: EQUIL
   38 Reading EQUIL    in file /media/sf_D_DRIVE/Linux/OPM/spe9_cp_edit/equil.txt, line 1

Warning: Expected the kewyord: EQLDIMS to infer the number of records in: RSVD   
```
Your input deck is very hard to read with having a separate include file for each keyword, as well as the DX  entries for each cell, why not use 9000*3.4116656046679076e+01 instead. 

If you cleanup the deck into one file and simplify the input then I'll take a look at it again and try to get to the bottom of this.

**Comment by OPMUSER** [comment]
@andrescodas If you do simplify your input deck do it on the original version not the DEPTH modified version.

**Comment by andrescodas** [comment]
Hi @OPMUSER.  Sorry for the inconvenience created with a lot of files.  As you could imagine, the deck is generated with an automated process.  Although I can fix the EQLDIMS and WELLDIMS issue, modifying the automated deck translator will not be an easy task.  Particularly, I don't know when I will have a deck that is simple enough for your analysis.  Indeed, the art of debugging is tightly related to finding the minimum failing example.

I could help if you point out exactly what are the simplifications you require.  Particularly, given that you use DEPTH PORV TRANX TRANY TRANZ in EDIT, whatever nonzero constant you place in DX DY DZ TOPS PERMX PERMY PERMZ PORO is irrelevant.  So, that could be one simplification to make...

**Comment by OPMUSER** [comment]
Okay I re-organized the input decks to see what is going on. The results are definitely different. And it looks like OPM Flow has the same values for the TOPS and the DEPTHS arrays, which you can see below - and that will cause issues.

![image](https://user-images.githubusercontent.com/17958927/71250650-294b0580-235b-11ea-96bd-4b50ebcf7892.png)

So I deleted the depth arrays and re-ran. As you can see below the below, the results are much more closer, but still not the exactly the same:
![SPE9-CP-OPM1910-R02](https://user-images.githubusercontent.com/17958927/71251691-066e2080-235e-11ea-9fcb-12b98d20c27a.PNG)

Could this be a result of time stepping differences ?
![image](https://user-images.githubusercontent.com/17958927/71252514-2a326600-2360-11ea-8e95-611fb089494e.png)

Does that help?

**Comment by OPMUSER** [comment]
@andrescodas when you use DX, DY and DZ in the commercial simulator it calculates the transmissiblities using the OLDTRAN option. OPM Flow does not have this option so it will always use the NEWTRAN option. So depending on where (or how) you derived the transmissibilities for the EDIT section, they will either be OLDTRAN or NEWTRAN based. For example, the following two plots, shows the commercial simulator reference case based on corner point geometry (NEWTRAN),  OPM Flow (SPE09-BASE01-OPM1910-R01 - DX, DY, DZ with NEWTRAN), and SPE09-BASE02-OPM1910-R01 (corner point geometry - NEWTRAN).

![SPE9_CP](https://user-images.githubusercontent.com/17958927/71353772-087dec80-25b5-11ea-90ec-e15e88103fd5.PNG)
![SPE9_CP-01](https://user-images.githubusercontent.com/17958927/71354107-e769cb80-25b5-11ea-9b7d-8b0291f7d63c.PNG)

Notice how the commercial simulator and OPM Flow NEWTRAN cases are identical, but the OPM Flow SPE09-BASE01-OPM1910-R01 - DX, DY, DZ with NEWTRAN case is different. 

I think this is the source of the miss-matching of results you are observing.

In summary, for bench marking and for comparison purposes always use corner point geometry in OPM Flow.

**Comment by andrescodas** [comment]
Hi @OPMUSER 

Thanks for pushing forward

> Could this be a result of time stepping differences ?

When time allows I will try to check this.  I guess I'll need to simulate with the commercial simulator and then input the same time steps to OPM-flow.  It will take me some time because I'll have to go through the manuals to check how to do it exactly.

> So depending on where (or how) you derived the transmissibilities for the EDIT section, they will either be OLDTRAN or NEWTRAN based.

I provide all the transmissibilities explicitly (and replace any value previously computed) by means of TRANX TRANY TRANZ in EDIT.  Further, I use the 8th record of COMPDAT to provide the transmissilibity explicitly for the well completions. Thus, regardless of OLDTRAN or NEWTRAN, or how the geometry is specified, the transmissibilities must be the same.

**Comment by alfbr** [comment]
Just do the easy thing, set a low maximum timestep in both. Flow supports doing that from the command line parameter --solver-max-time-step-in-days, Eclipse in the TUNING keyword. Time stepping can indeed make it difficult to compare results.

**Comment by andrescodas** [comment]
I repeated the experiments forcing the timesteps to 1 day.  The results are still different and consistent with the previous simulations here https://github.com/OPM/opm-simulators/issues/2183#issuecomment-564132954:

![Oil_Production_Rate](https://user-images.githubusercontent.com/164282/71414396-897cc880-2635-11ea-8918-05272187c1d2.png)

**Comment by andrescodas** [comment]
I found that the transmissibilities are not consistent in the .INIT file.  I checked it with the following MRST code:

`mrstModule add deckformat`
`init_ecl  = readEclipseOutputFileUnFmt('SPE9_EDIT_NO_DEPTH_ECLIPSE.INIT');`
`init_opm  = readEclipseOutputFileUnFmt('SPE9_EDIT_NO_DEPTH_OPM.INIT');`

`ix_opm = init_opm.TRANX.values~=0;`
`ix_ecl = init_ecl.TRANX.values~=0;`

`fprintf('Nonzero TRANX opm %d\n',sum(ix_opm))`
`fprintf('Nonzero TRANX ecl %d\n',sum(ix_ecl))`

`assert(norm(init_ecl.TRANX.values(ix_opm)-init_opm.TRANX.values(ix_opm))==0);`

Simulator inputs:
[spe9.zip](https://github.com/OPM/opm-simulators/files/3998565/spe9.zip)


Simulator INIT output
[spe9_edit_no_depth_init.zip](https://github.com/OPM/opm-simulators/files/3998559/spe9_edit_no_depth_init.zip)

**Comment by OPMUSER** [comment]
Okay there appears to be a bug in TRANX array. 

The following plots show the transmissibilty differences between the two cases, only TRANX shows a difference.
![image](https://user-images.githubusercontent.com/17958927/71431166-16458600-270b-11ea-9ea7-2b08ca906f52.png)
![image](https://user-images.githubusercontent.com/17958927/71431170-1fceee00-270b-11ea-96ac-ea7917e96e76.png)
![image](https://user-images.githubusercontent.com/17958927/71431178-28bfbf80-270b-11ea-8482-4864c1419169.png)

**Comment by OPMUSER** [comment]
If one plots the TRANX on a log scale we gets this:

![image](https://user-images.githubusercontent.com/17958927/71497820-de6a4a00-2894-11ea-8c0e-a89e6543dcee.png)
And then filtering on non-zeros and switching off the grid cells, we get this:

![image](https://user-images.githubusercontent.com/17958927/71498294-bda2f400-2896-11ea-9b81-3b46e16e8188.png)

This is definitely an OLDTRAN/NEWTRAN issue, as  OPM Flow does not have an OLDTRAN option. To prove this I set the TOPS keyword to a constant value so we have a flat grid and thus OLDTRAN and NEWTRAN will give us the same TRANXs, so now we have:
![image](https://user-images.githubusercontent.com/17958927/71500644-d57f7580-28a0-11ea-8224-f3a25acc09d2.png)

And we get identical results :+1: 
![image](https://user-images.githubusercontent.com/17958927/71500672-f6e06180-28a0-11ea-897f-d5b5fabd7065.png)

**Comment by OPMUSER** [comment]
Hit the wrong key before I finished - I have updated my previous comment with an additional figure and text.

**Comment by bska** [comment]
I **suppose** you could provide the simulation graph as a one-D grid with NNCs, but this gets cumbersome quickly. As for OLDTRAN, that is indeed low priority.

**Comment by andrescodas** [comment]
There are two separate discussion lines to be addressed here:

1. OLDTRAN/NEWTRAN
2. TRANX/Y/X and NNC

Let me address both of them here:

1. OLDTRAN/NEWTRAN:

-->  I included NEWTRAN on the input of both simulators.  If the difference between OLDTRAN/NEWTRAN  is the source of the issue, then the simulations should match. See the result:
![Oil_Production_Rate_NEWTRAN](https://user-images.githubusercontent.com/164282/71516252-a65c0a80-2886-11ea-918f-de0e016d5672.png)

Thus, since the difference remain, this experiment rules out the hypothesis that OLDTRAN/NEWTRAN is behind the issue.

2. TRANX/Y/X and NNC

The flow documentation manual reads:

![Screenshot from 2019-12-27 09-01-33](https://user-images.githubusercontent.com/164282/71516407-9264d880-2887-11ea-8274-c56253cdda54.png)

You say:
>TRANX/Y/X are the transmissibilities between grid faces only, not between NNCs

I understand from the manual (both from OPM and Eclipse) that TRANX resets the transmissibility between the blocks (I, J, K) and (I+1, J, K), regardless if their faces are geometrically apart of each other.  At least there is no note indicating that we should make sure the faces are indeed overlapping.

I see your point, that since the cells don't have overlapping faces they could be thought as 'Non-Neighboring Connections', but I think 'Non-Neighboring Connections' here refers to the indexing of the cells, and not the geometry.  Thus, non-neighbor connections are (I, J, K) - (I+2, J, K), and (I, J, K) - (I+1, J, K), are neighbors, regardless of the actual geometry

**Comment by andrescodas** [comment]
Hi @bska 

> I suppose you could provide the simulation graph as a one-D grid with NNCs, but this gets cumbersome quickly. As for OLDTRAN, that is indeed low priority.

I tried this before and although it works, it is not efficient because the linear solvers depend on cells ordering to compute pre-conditioners

**Comment by OPMUSER** [comment]
@andrescodas ,

I don't think your test case of using OLDTRAN in the commercial simulator is valid, as the commercial simulator manual clearly states:

`Although it is normal to use OLDTRAN with DX, DY, DZ, TOPS etc. input, and NEWTRAN with COORD/ZCORN, all types of transmissibility calculation are available with both types of grid specification, as the simulator holds grid data internally in both forms. Corner points are obtained from DX/DY/DZ/TOPS data by assuming flat blocks, and DX/DY/DZ values are obtained from corner points by evaluation of the distances between cell face center points. However, use of NEWTRAN with DX/DY/DZ/TOPS input may well produce large numbers of spurious fault connections, as the assumed flat grid blocks in neighboring columns will generally overlap, and this combination of options is not recommended.`

Comparing TRANX on a log scale using NEWTRAN in both simulators gives this:

![image](https://user-images.githubusercontent.com/17958927/71538540-bc8dc780-2967-11ea-91b4-9d3e26a33141.png)

Clearly the commercial simulator is wrong here one cannot have non-zero transmissibilities with non-connecting blocks. Which is why I generated the constant TOPS runs mentioned above.

Now if we compare the commercial simulators OLDTRAN and NEWTRAN cases we get this:

![image](https://user-images.githubusercontent.com/17958927/71538588-a46a7800-2968-11ea-867b-3e281ff7e433.png)

Looks fairly similar, but Like I said the commercial simulator clearly warns the user not to use this combination.

To summarize, if you wish to use OPM Flow then you should us corner-point geometry in order to obtain similar results to the commercial simulator. There are numerous test cases using corner-point geometry that give identical results between the two simulators. Alternatively, you could implement an OLDTRAN option.

**Comment by andrescodas** [comment]
Hi @OPMUSER 

Thanks for bearing with me. I followed your advice and constructed a corner-point grid with the same geometry in this discussion.  See the case attached

The differences in TRANX persist,  see the attachment.  While the commercial simulator sets the values of TRANX regardless if the logical consecutive grid-blocks are touching or not, OPM doesn't.

In my opinion the commercial simulator has the intended behavior because it is more intuitive once the user forces a non-zero transmissibility with the TRANX keyword.
[CP.zip](https://github.com/OPM/opm-simulators/files/4007285/CP.zip)

**Comment by OPMUSER** [comment]
@andrescodas okay I looked at this again and I agree there is definitely something wrong here when you load the TRAN values in the EDIT section. It looks like the bug only effects the TRANX array, as the TRANY and TRANZ arrays are identical between the two simulators. The figure below shows the TRANX array for the two simulators. Notice how OPM Flow has zero's for most of the array except the last layer. In fact the last layer has the same values as the commercial simulator.

![image](https://user-images.githubusercontent.com/17958927/71566601-cc3a1700-2af3-11ea-9c23-302ee17e55c5.png)

I tried using the BOX and ENDBOX keywords around all the TRAN include files, but I also got the same result. I then tried repeating the TRANX keyword after the TRANZ data, but again I got the same result. I also checked the OPM Flow oil rates in case this was a visualization issue; however, all the OPM Flow cases were identical, meaning it is a TRANX bug.

So yes there is a bug when entering the TRANX keyword in the edit section and at the moment I don't have a work around. Hopefully one of the developers will take a look at this and get to the bottom of this.
